contracts/account/extensions/ERC7579Modules/ERC7579SocialRecoveryExecutor.sol

```diff
@@ -0,0 +1,453 @@
+    /// @return The EIP-712 digest for starting recovery
+    function _getStartRecoveryDigest(
+        address account,
+        bytes calldata executionCalldata
+    ) internal view returns (bytes32) {
+        bytes32 structHash = keccak256(
+            abi.encode(START_RECOVERY_TYPEHASH, account, keccak256(executionCalldata), nonces(account))
+        );
+        return _hashTypedDataV4(structHash);
+    }
+
+    /// @notice EIP-712 digest for cancelling recovery
+    /// @param account The ERC-7579 Account to cancel recovery for
+    /// @param nonce The nonce of the account
+    /// @return The EIP-712 digest for cancelling recovery
+    function _getCancelRecoveryDigest(address account, uint256 nonce) internal view returns (bytes32) {
+        bytes32 structHash = keccak256(abi.encode(CANCEL_RECOVERY_TYPEHASH, account, nonce));
+        return _hashTypedDataV4(structHash);
+    }
+
+    /// @notice Verifies multiple guardian signatures for a given digest
+    /// @dev Ensures signatures are unique, and threshold is met
+    /// @param account The account the signatures are for
+    /// @param guardianSignatures Array of guardian signatures sorted by signer address in ascending order
+    /// @param digest The digest to verify the signatures against
+    function _validateGuardianSignatures(
+        address account,
+        GuardianSignature[] calldata guardianSignatures,
+        bytes32 digest
+    ) internal view virtual {
+        // bound `for` cycle
+        if (guardianSignatures.length > maxGuardians()) {
+            revert TooManyGuardianSignatures();
+        }
+        if (guardianSignatures.length < _recoveryConfigs[account].threshold) {
+            revert ThresholdNotMet();
+        }
+
+        address lastSigner = address(0);
+        for (uint256 i = 0; i < guardianSignatures.length; i++) {
+            address signer = guardianSignatures[i].signer;
+            if (
+                !isGuardian(account, signer) ||
+                !SignatureChecker.isValidSignatureNow(signer, digest, guardianSignatures[i].signature)
+            ) {
+                revert InvalidGuardianSignature();
+            }
+            if (uint160(signer) <= uint160(lastSigner)) {
+                revert DuplicatedOrUnsortedGuardianSignatures();
+            }
+            lastSigner = signer;
+        }
+    }
+
+    /// @notice Cancels an ongoing recovery process
+    /// @param account The ERC-7579 Account to cancel recovery for
+    function _cancelRecovery(address account) internal virtual {
+        _recoveryConfigs[account].pendingExecutionHash = bytes32(0);
+        _recoveryConfigs[account].recoveryStart = 0;
+        emit RecoveryCancelled(account);
+    }
+
+    /// @notice Adds a new guardian to the account's recovery configuration
+    /// @param account The ERC-7579 Account to add the guardian to
+    /// @param guardian Address of the new guardian
+    function _addGuardian(address account, address guardian) internal virtual {
```

Comment:
I think the signature of both the `_addGuardian` and `_removeGuardian` functions must allow batch adding/removing guardians natively. It'd be more elegant but it's also opinionated since the account can batch operations. It's just that batching operations at the account level may require extra work than just calling a batched function directly

```solidity
for (uint256 i = 0; i < signers.length; i++) {
    if (!_associatedSigners[account].add(guardians[i]))
        revert AlreadyGuardian(account, guardians[i]);
}
emit GuardiansAdded(account, guardians);
```